File: WebClient SDK/src/default/ui/krProcess/krSecondaryProcessUIExtension.ts
import { CardUIExtension, ICardModel, ICardUIExtensionContext } from 'tessa/ui/cards';
import { DotNetType, Visibility } from 'tessa/platform';
import { CardRowState } from 'tessa/cards';
import { ConditionsUIContext } from 'tessa/ui/conditions';

export class KrSecondaryProcessUIExtension extends CardUIExtension {
  private _model: ICardModel;

  private _pureProcess = {
    id: 0,
    name: '$KrSecondaryProcess_Mode_PureProcess'
  };

  private _button = {
    id: 1,
    name: '$KrSecondaryProcess_Mode_Button'
  };

  private _action = {
    id: 2,
    name: '$KrSecondaryProcess_Mode_Action'
  };

  private _dispose: Function | null;

  public shouldExecute(context: ICardUIExtensionContext) {
    return context.card.typeId === '61420fa1-cc1f-47cb-b0bb-4ea8ee77f51a';
  }

  public initialized(context: ICardUIExtensionContext) {
    this._model = context.model;

    this.initializeConditions();

    const currentMode = this.getCurrentMode();
    this.updateVisibility(currentMode);
    if (currentMode === this._pureProcess.id) {
      this.updateCheckRestrictions();
    }

    this._dispose = context.card.sections
      .get('KrSecondaryProcesses')!
      .fields.fieldChanged.addWithDispose(e => {
        switch (e.fieldName) {
          case 'ModeID':
            this.updateVisibility(this.getCurrentMode());
            break;

          case 'AllowClientSideLaunch':
            this.updateVisibiltyForPureProcessMode();
            break;

          case 'CheckRecalcRestrictions':
            this.updateCheckRestrictions();
            break;
        }
      });
  }

  public finalized() {
    if (this._dispose) {
      this._dispose();
      this._dispose = null;
    }
  }

  private updateVisibility(currentMode: number) {
    const getVisibility = (allowedMode: number, allowedMode2: number = -2) => {
      return allowedMode === currentMode || allowedMode2 === currentMode
        ? Visibility.Visible
        : Visibility.Collapsed;
    };

    const blocks = this._model.blocks;
    blocks.get('PureProcessParametersBlock')!.blockVisibility = getVisibility(this._pureProcess.id);
    blocks.get('TileParametersBlock')!.blockVisibility = getVisibility(this._button.id);
    blocks.get('ActionParametersBlock')!.blockVisibility = getVisibility(this._action.id);
    blocks.get('VisibilityScriptsBlock')!.blockVisibility = getVisibility(this._button.id);
    blocks.get('ConditionsTable')!.blockVisibility = getVisibility(this._button.id);

    blocks.get('RestictionsBlock')!.blockVisibility = Visibility.Visible;
    blocks.get('ExecutionAccessDeniedBlock')!.blockVisibility = Visibility.Visible;
    blocks.get('ExecutionScriptsBlock')!.blockVisibility = Visibility.Visible;

    if (currentMode === this._pureProcess.id) {
      this.updateVisibiltyForPureProcessMode();
    }
  }

  private updateVisibiltyForPureProcessMode() {
    const card = this._model.card;
    const sec = card.sections.get('KrSecondaryProcesses')!;
    const allowClientSideLaunch = sec.fields.tryGet('AllowClientSideLaunch') || false;
    const checkRecalcControl = this._model.blocks
      .get('PureProcessParametersBlock')!
      .controls.find(x => x.name === 'CheckRecalcRestrictionsCheckbox')!;

    checkRecalcControl.isReadOnly = allowClientSideLaunch;
    const checkRecalcRestrictions = sec.fields.tryGet('CheckRecalcRestrictions') || false;
    if (allowClientSideLaunch && !checkRecalcRestrictions) {
      sec.fields.set('CheckRecalcRestrictions', true, DotNetType.Boolean);
    } else if (!allowClientSideLaunch && !checkRecalcRestrictions) {
      this.updateCheckRestrictions();
    }
  }

  private updateCheckRestrictions() {
    const blocks = this._model.blocks;
    const card = this._model.card;
    const checkRecalcRestrictions =
      card.sections.get('KrSecondaryProcesses')!.fields.tryGet('CheckRecalcRestrictions') || false;
    const visibilityForRestrictionFields = checkRecalcRestrictions
      ? Visibility.Visible
      : Visibility.Collapsed;
    blocks.get('RestictionsBlock')!.blockVisibility = visibilityForRestrictionFields;
    blocks.get('ExecutionAccessDeniedBlock')!.blockVisibility = visibilityForRestrictionFields;
    blocks.get('ExecutionScriptsBlock')!.blockVisibility = visibilityForRestrictionFields;

    if (!checkRecalcRestrictions) {
      const sec = card.sections.get('KrSecondaryProcesses')!;
      sec.fields.set('ExecutionAccessDeniedMessage', null);
      sec.fields.set('ExecutionSqlCondition', null);
      sec.fields.set('ExecutionSourceCondition', null);

      const clear = (name: string) => {
        const rows = card.sections.get(name)!.rows;
        const removeRows = rows.filter(x => x.state === CardRowState.Inserted);
        for (let row of removeRows) {
          rows.remove(row);
        }
        for (let row of rows) {
          row.state = CardRowState.Deleted;
        }
      };

      clear('KrStageDocStates');
      clear('KrStageTypes');
      clear('KrStageRoles');
      clear('KrSecondaryProcessRoles');
    }
  }

  private getCurrentMode(): number {
    const modeId = this._model.card.sections.get('KrSecondaryProcesses')!.fields.tryGet('ModeID');
    // tslint:disable-next-line:triple-equals
    return modeId == undefined ? -1 : modeId;
  }

  private initializeConditions() {
    const conditionContext = new ConditionsUIContext();
    conditionContext.initialize(this._model);
  }
}

File: Source/Extensions/Tessa.Extensions.Default.Client/UI/KrProcess/KrSecondaryProcessUIExtension.cs
using System.Linq;
using System.Threading.Tasks;
using System.Windows;
using Tessa.Cards;
using Tessa.Extensions.Default.Shared.Workflow.KrProcess;
using Tessa.Extensions.Platform.Client.UI;
using Tessa.Platform;
using Tessa.Platform.Conditions;
using Tessa.Platform.Placeholders;
using Tessa.Platform.Runtime;
using Tessa.Platform.Storage;
using Tessa.UI.Cards;
using Unity;

namespace Tessa.Extensions.Default.Client.UI.KrProcess
{
    public sealed class KrSecondaryProcessUIExtension : CardUIExtension
    {
        #region Fields

        private static readonly int pureProcess = KrConstants.KrSecondaryProcessModes.PureProcess.ID;
        private static readonly int button = KrConstants.KrSecondaryProcessModes.Button.ID;
        private static readonly int action = KrConstants.KrSecondaryProcessModes.Action.ID;

        private ICardModel model;
        private CardSection secondaryProcessesSection;

        private readonly ISession session;
        private readonly IUnityContainer unityContainer;
        private readonly ICardRepairManager cardRepairManager;
        private readonly ICardMetadata cardMetadata;
        private readonly IConditionTypesProvider typesProvider;
        private readonly IPlaceholderManager placeholderManager;

        #endregion

        #region Constructors

        public KrSecondaryProcessUIExtension(
            ISession session,
            IUnityContainer unityContainer,
            ICardRepairManager cardRepairManager,
            ICardMetadata cardMetadata,
            IConditionTypesProvider typesProvider,
            IPlaceholderManager placeholderManager)
        {
            this.session = session;
            this.unityContainer = unityContainer;
            this.cardRepairManager = cardRepairManager;
            this.cardMetadata = cardMetadata;
            this.typesProvider = typesProvider;
            this.placeholderManager = placeholderManager;
        }

    #endregion

    #region Base Overrides

    /// <inheritdoc/>
    public override async Task Initialized(ICardUIExtensionContext context)
        {
            this.model = context.Model;

            this.InitializeConditions();

            var currentMode = this.GetCurrentMode();
            this.UpdateVisibility(currentMode);
            if (currentMode == pureProcess)
            {
                this.UpdateCheckRestrictions();
            }

            this.secondaryProcessesSection = context.Card
                .Sections[KrConstants.KrSecondaryProcesses.Name];
            this.secondaryProcessesSection.FieldChanged += this.FieldChanged;
        }

        /// <inheritdoc/>
        public override async Task Finalized(ICardUIExtensionContext context)
        {
            if (this.secondaryProcessesSection is not null)
            {
                this.secondaryProcessesSection.FieldChanged -= this.FieldChanged;
                this.secondaryProcessesSection = default;
            }

            this.model = null;
        }

        #endregion

        #region Private Methods

        private void FieldChanged(object sender, CardFieldChangedEventArgs e)
        {
            switch (e.FieldName)
            {
                case KrConstants.KrSecondaryProcesses.ModeID:
                    this.UpdateVisibility(this.GetCurrentMode());
                    break;
                case KrConstants.KrSecondaryProcesses.AllowClientSideLaunch:
                    this.UpdateVisibiltyForPureProcessMode();
                    break;
                case KrConstants.KrSecondaryProcesses.CheckRecalcRestrictions:
                    this.UpdateCheckRestrictions();
                    break;
            }
        }

        private void UpdateVisibility(int currentMode)
        {
            Visibility GetVisibility(int allowedMode, int allowedMode2 = -2) => 
                allowedMode == currentMode || allowedMode2 == currentMode
                ? Visibility.Visible
                : Visibility.Collapsed;

            var blocks = this.model.Blocks;
            blocks[KrConstants.Ui.PureProcessParametersBlock].BlockVisibility = GetVisibility(pureProcess);
            blocks[KrConstants.Ui.TileParametersBlock].BlockVisibility = GetVisibility(button);
            blocks[KrConstants.Ui.ActionParametersBlock].BlockVisibility = GetVisibility(action);
            blocks[KrConstants.Ui.VisibilityScriptsBlock].BlockVisibility = GetVisibility(button);
            blocks[ConditionsUIContext.ConditionControlName].BlockVisibility = GetVisibility(button);

            blocks[KrConstants.Ui.RestictionsBlock].BlockVisibility = Visibility.Visible;
            blocks[KrConstants.Ui.ExecutionAccessDeniedBlock].BlockVisibility = Visibility.Visible;
            blocks[KrConstants.Ui.ExecutionScriptsBlock].BlockVisibility = Visibility.Visible;

            if (currentMode == pureProcess)
            {
                this.UpdateVisibiltyForPureProcessMode();
            }
        }

        private void UpdateVisibiltyForPureProcessMode()
        {
            var card = this.model.Card;
            var sec = card.Sections[KrConstants.KrSecondaryProcesses.Name];
            var allowClientSideLaunch =
                sec.Fields.TryGet<bool?>(KrConstants.KrSecondaryProcesses.AllowClientSideLaunch) ?? false;
            var checkRecalcControl = this.model
                .Blocks[KrConstants.Ui.PureProcessParametersBlock]
                .Controls
                .First(p => p.Name == KrConstants.Ui.CheckRecalcRestrictionsCheckbox);

            checkRecalcControl.IsReadOnly = allowClientSideLaunch;
            var checkRecalcRestrictions =
                sec.Fields.TryGet<bool?>(KrConstants.KrSecondaryProcesses.CheckRecalcRestrictions) ?? false;
            if (allowClientSideLaunch 
                && !checkRecalcRestrictions)
            {
                sec.Fields[KrConstants.KrSecondaryProcesses.CheckRecalcRestrictions] = BooleanBoxes.True;
            }
            else if (!allowClientSideLaunch
                && !checkRecalcRestrictions)
            {
                this.UpdateCheckRestrictions();
            }
        }

        private void UpdateCheckRestrictions()
        {
            var blocks = this.model.Blocks;
            var card = this.model.Card;
            var checkRecalcRestrictions = card
                .Sections[KrConstants.KrSecondaryProcesses.Name]
                .Fields
                .TryGet<bool?>(KrConstants.KrSecondaryProcesses.CheckRecalcRestrictions) ?? false;
            var visibilityForRestrictionFields = checkRecalcRestrictions
                ? Visibility.Visible
                : Visibility.Collapsed;
            blocks[KrConstants.Ui.RestictionsBlock].BlockVisibility = visibilityForRestrictionFields;
            blocks[KrConstants.Ui.ExecutionAccessDeniedBlock].BlockVisibility = visibilityForRestrictionFields;
            blocks[KrConstants.Ui.ExecutionScriptsBlock].BlockVisibility = visibilityForRestrictionFields;

            if (!checkRecalcRestrictions)
            {
                var sec = card.Sections[KrConstants.KrSecondaryProcesses.Name];
                sec.Fields[KrConstants.KrSecondaryProcesses.ExecutionAccessDeniedMessage] = null;
                sec.Fields[KrConstants.KrSecondaryProcesses.ExecutionSqlCondition] = null;
                sec.Fields[KrConstants.KrSecondaryProcesses.ExecutionSourceCondition] = null;

                Clear(KrConstants.KrStageDocStates.Name);
                Clear(KrConstants.KrStageTypes.Name);
                Clear(KrConstants.KrStageRoles.Name);
                Clear(KrConstants.KrSecondaryProcessRoles.Name);
                void Clear(string name)
                {
                    var rows = card.Sections[name].Rows;
                    rows.RemoveAll(p => p.State == CardRowState.Inserted);
                    foreach (var row in rows)
                    {
                        row.State = CardRowState.Deleted;
                    }
                }
            }
        }

        private int GetCurrentMode() => this.model.Card
            .Sections[KrConstants.KrSecondaryProcesses.Name]
            .Fields
            .TryGet<int?>(KrConstants.KrSecondaryProcesses.ModeID) ?? -1;

        private void InitializeConditions()
        {
            var conditionContext = new ConditionsUIContext(
                this.session,
                this.unityContainer,
                this.cardRepairManager,
                this.cardMetadata,
                this.typesProvider,
                this.placeholderManager);

            conditionContext.Initialize(this.model);
        }

        #endregion
    }
}